practice6.1/dispatcher.cpp

```diff
@@ -0,0 +1,129 @@
+#include <iostream>
+#include <vector>
+#include <string>
+#include <functional>
+#include <thread>
+#include <mutex>
+#include <random>
+#include <chrono>
+#include <print>
+
+class GreenMsg {
+public:
+    std::string text;
+    int counter;
+
+    GreenMsg(const std::string& t, int c) : text(t), counter(c) {}
+};
+
+class BlueMsg {
+public:
+    double value1;
+    double value2;
+
+    BlueMsg(double v1, double v2) : value1(v1), value2(v2) {}
+};
+
+class OrangeMsg {
+public:
+    std::string text1;
+    std::string text2;
+    int number;
+    double value;
+
+    OrangeMsg(const std::string& t1, const std::string& t2, int n, double v) 
+        : text1(t1), text2(t2), number(n), value(v) {}
+};
+
+class MsgDispatcher {
+public:
+    void subscribeToGreenMsg(const std::function<void(const GreenMsg&)>& subscriber) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        GreenMsgSubscribers_.push_back(subscriber);
+    }
+
+    void subscribeToBlueMsg(const std::function<void(const BlueMsg&)>& subscriber) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        BlueMsgSubscribers_.push_back(subscriber);
+    }
+
+    void subscribeToOrangeMsg(const std::function<void(const OrangeMsg&)>& subscriber) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        OrangeMsgSubscribers_.push_back(subscriber);
+    }
+
+    void publishGreenMsg(const GreenMsg& message) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        for (const auto& subscriber : GreenMsgSubscribers_) {
+            subscriber(message);
```

Comment:
Your function performs not only publish but also handling of messages. And we don't know how much time it take to run the callback. Publish is expected to be almost instant operation when we put message to the queue and continue execution while message will be dispatched by a separate thread.